Archives/ClmFile.cpp

```diff
@@ -405,24 +390,57 @@ namespace Archives
 		return strippedExtensions;
 	}
 
+	WaveFormatEx ClmFile::CreateDefaultWaveFormat() {
+		WaveFormatEx waveFormat;
+
+		waveFormat.wFormatTag = 1; // WAVE_FORMAT_PCM
+		waveFormat.nChannels = 1; // mono
+		waveFormat.nSamplesPerSec = 22050; // 22.05KHz
+		waveFormat.nAvgBytesPerSec = 44100; // nSamplesPerSec *nBlockAlign
+		waveFormat.nBlockAlign = 2; // 2 bytes/sample = nChannels * wBitsPerSample / 8
+		waveFormat.wBitsPerSample = 16;
+		waveFormat.cbSize = 0;
+
+		return waveFormat;
+	}
+
 
 	// A null terminator (\0) is automatically assigned to the end of the string when placing it within the std::array
 	const std::array<char, 32> standardFileVersion { "OP2 Clump File Version 1.0\x01A\0\0\0\0" };
 	const std::array<char, 6> standardUnknown { 0, 0, 0, 0, 1, 0 };
 
 	ClmFile::ClmHeader::ClmHeader(WaveFormatEx waveFormat, uint32_t packedFilesCount) {
-		std::memcpy(fileVersion.data(), standardFileVersion.data(), fileVersion.size());
-		std::memcpy(unknown.data(), standardUnknown.data(), unknown.size());
+		fileVersion = standardFileVersion;
+		unknown = standardUnknown;
 
 		this->waveFormat = waveFormat;
 		this->packedFilesCount = packedFilesCount;
 	}
 
 	bool ClmFile::ClmHeader::CheckFileVersion() const {
-		return std::memcmp(fileVersion.data(), standardFileVersion.data(), fileVersion.size()) == 0;
+		return fileVersion == standardFileVersion;
 	};
 
 	bool ClmFile::ClmHeader::CheckUnknown() const {
-		return std::memcmp(unknown.data(), standardUnknown.data(), unknown.size()) == 0;
+		return unknown == standardUnknown;
+	}
+
+	void ClmFile::ClmHeader::VerifyFileVersion() const {
+		if (!CheckFileVersion()) {
+			throw std::runtime_error("CLM file version is incorrect.");
+		}
+	}
+
+	void ClmFile::ClmHeader::VerifyUnknown() const {
+		if (!CheckUnknown()) {
+			throw std::runtime_error("Unknown field in CLM header is incorrect.");
+		}
+	}
+
+	std::string ClmFile::IndexEntry::GetFileName() const {
+		char fileNameOut[9];
+		strncpy_s(fileNameOut, fileName.data(), 8);
+
+		return std::string(fileNameOut);
```

Comment:
This looks correct, and doesn't leak memory, though it does do a double string copy. There's probably a way to create a `std::string` with 9 characters of buffer space, and copy the `std::array` directly into that and then null terminate it.

This StackOverflow question on [Copying a std::array to a std::string](https://stackoverflow.com/questions/10900345/stdarraychar-n-to-stdstring) may help. It probably handles null termination for you.

Reply:
I was having difficulty getting from std::array to std::string with a proper null terminator without performing a double copy. I'll check the link you provided out and give it another shot as I don't like my current solution either.

Reply:
I did some testing, and this seems like it should work:
```
std::string ClmFile::IndexEntry::GetFileName() const {
    return std::string(fileName.data(), fileName.size());
}
```
That should grab 8 chars from the buffer, put them into a `std::string` object, and null terminate it (max 9 chars in the string buffer, including the null terminator). If the string is shorter, such that it doesn't take up the full 8 characters, the earlier null will terminate the string.

It uses the `string` constructor:
```
string(const char* s, size_t n);
```